### PR TITLE
Video cache

### PR DESCRIPTION
**Pull request by GrahamGoudeau**

### CODE REVIEW

**Comment by GrahamGoudeau** [comment]
Closes #104

**Comment by ethanpailes** [comment]
This looks really clean.

It looks like the video cache only works in-memory, and most of the value-added from having a history page is that we don't have to re-download videos after restarting the app. The video files are going to be there on disk anyway, so it is worth listing them.

If you could add the ability to serialize the video metadata and access time to some location in the CONFIG.APP_HOME_DIR, and then read it on startup, that would be ideal.

File: app/client/ts/video-cache.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    static getInstance() {
+        if (VideoCache.INSTANCE == null) {
+            VideoCache.INSTANCE = new VideoCache();
+        }
+        return VideoCache.INSTANCE;
+    }
+
+    async getFromXorName(xorName: string): Promise<Video> {
+        if (this.cache[xorName] == null) {
+            if (this.numEntries + 1 > CONFIG.MAX_CACHE_SIZE) {
```
Comment (ethanpailes) [inline app/client/ts/video-cache.ts:74]: I like the garbage-collection bit.

**Review by ethanpailes — CHANGES_REQUESTED** [review]
See my comment in the main thread about serializing the cache to disk.

**Comment by nicholasyan** [comment]
With the changes @ethanpailes was talking about, looks good to me.

**Comment by ethanpailes** [comment]
I just merged #114 BTW

File: app/client/ts/video-model.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ class UnsupportedVideoFormatError extends Error {
     }
 }
 
-export default class Video implements Drop {
+export class Video implements Drop {
```
Comment (ethanpailes) [inline app/client/ts/video-model.ts]: Video really is the main thing this module exports. You can have both a default export and standard exports.

File: app/client/ts/video-cache.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            this.numEntries += 1;
+        } else {
+            this.cache[xorName].accessTime = new Date();
+        }
+
+        console.log('cache state:', this.cache);
+        return this.cache[xorName].video;
+    }
+
+    stateToString(): string {
```
Comment (ethanpailes) [inline app/client/ts/video-cache.ts]: What is going on here?

File: app/client/ts/video-cache.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        //const ret = Object.keys(this.cache).map((xorName: string) => return {
+        return JSON.stringify({
+            xorName: '',
+            name: ''
+        });
+    }
+
+    private static async readFromDisk(longName): Promise<Maybe<CacheStateStringy>> {
```
Comment (ethanpailes) [inline app/client/ts/video-cache.ts]: What does Nothing represent that is different than an empty CacheStateStringy object? A comment explaining might be a little nice. (non merge blocking).

File: app/client/ts/video-cache.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        const homeDir: DnsHomeDirectory = await safeClient.dns.getHomeDirectory(longName, CONFIG.SERVICE_NAME);
```
Comment (ethanpailes) [inline app/client/ts/video-cache.ts]: I may be misunderstanding what this function is supposed to do, but why is it touching the SafeNET? The cache metadata should be stored locally to describe the files that actually exist on disk on your current machine. It does not make any sense to download a video on my linux box, update my cache on the SafeNET, then move to my mac and try to read one of the local files that I've downloaded onto my linux box.

File: app/client/ts/video-cache.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+        if (fileNames.indexOf(VideoCache.CACHE_STATE_FILENAME) === -1) {
+            return Maybe.nothing<CacheStateStringy>();
+        }
+
+        const dataFile: SafeFile = await safeClient.dns.getFile(longName, CONFIG.SERVICE_NAME, VideoCache.CACHE_STATE_FILENAME);
+        return Maybe.just<CacheStateStringy>(JSON.parse(dataFile.body.toString()));
+    }
+
+    async persistOnDisk() {
```
Comment (ethanpailes) [inline app/client/ts/video-cache.ts]: Same concern as above. I think this stuff should really be mucking about with files under the CONFIG.SAFE_APP_DIR directory.

**Review by ethanpailes — CHANGES_REQUESTED** [review]

**Comment by ethanpailes** [comment]
It doesn't look like this will get maintained.

**Pull request closed without merging**